java/aiservices/openai/src/main/java/com/microsoft/semantickernel/aiservices/openai/chatcompletion/OpenAIChatCompletion.java

```diff
@@ -82,170 +74,140 @@ public Mono<List<ChatMessageContent>> getChatMessageContentsAsync(ChatHistory ch
         @Nullable PromptExecutionSettings promptExecutionSettings, @Nullable Kernel kernel) {
 
         List<ChatRequestMessage> chatRequestMessages = getChatRequestMessages(chatHistory);
-        List<FunctionDefinition> functions = Collections.emptyList();
-        return internalChatMessageContentsAsync(chatRequestMessages, functions,
-            promptExecutionSettings);
-
+        return internalChatMessageContentsAsync(chatRequestMessages, promptExecutionSettings, kernel);
     }
 
     @Override
     public Mono<List<ChatMessageContent>> getChatMessageContentsAsync(String prompt,
         PromptExecutionSettings promptExecutionSettings, Kernel kernel) {
         ParsedPrompt parsedPrompt = XMLPromptParser.parse(prompt);
+
         return internalChatMessageContentsAsync(
             parsedPrompt.getChatRequestMessages(),
-            parsedPrompt.getFunctions(),
-            promptExecutionSettings);
-    }
-
-    private Mono<List<ChatMessageContent>> internalChatMessageContentsAsync(
-        List<ChatRequestMessage> chatRequestMessages,
-        List<FunctionDefinition> functions,
-        PromptExecutionSettings promptExecutionSettings) {
-        ChatCompletionsOptions options = getCompletionsOptions(this, chatRequestMessages, functions,
-            promptExecutionSettings);
-        Mono<List<ChatMessageContent>> results =
-            internalChatMessageContentsAsync(options);
-
-        return results.flatMap(list -> {
-            boolean makeSecondCall = false;
-            for (ChatMessageContent messageContent : list) {
-                if (messageContent.getAuthorRole() == AuthorRole.TOOL) {
-                    makeSecondCall = true;
-                    String content = messageContent.getContent();
-                    String id = messageContent.getModelId();
-                    ChatRequestToolMessage toolMessage = new ChatRequestToolMessage(content, id);
-                    chatRequestMessages.add(toolMessage);
-                }
-            }
-            if (makeSecondCall) {
-                return internalChatMessageContentsAsync(options);
-            }
-            return Mono.just(list);
-        });
+            promptExecutionSettings,
+            kernel);
     }
 
     private Mono<List<ChatMessageContent>> internalChatMessageContentsAsync(
-        ChatCompletionsOptions options) {
-
-        return client
-            .getChatCompletions(getModelId(), options)
-            .filter(choices -> choices.getChoices() != null && !choices.getChoices().isEmpty())
-            .map(this::accumulateResponses)
-            .map(responses ->
-                responses.stream()
-                    .map(ChatResponseCollector::toChatMessageContent)
-                    .collect(Collectors.toList())
+        List<ChatRequestMessage> messages,
+        PromptExecutionSettings settings,
+        Kernel kernel) {
+
+        List<FunctionDefinition> functions = new ArrayList<>();
+        if (kernel != null) {
```

Comment:
if passing a null kernel is valid here then the Kernel argument should be `@Nullable` but I am not sure if calling with a null kernel would be valid

Reply:
> if passing a null kernel is valid here then the Kernel argument should be `@Nullable` but I am not sure if calling with a null kernel would be valid

Sure. Why not? Look at Example17_ChatGPT for example.